examples/qoi.cpp

```diff
@@ -0,0 +1,205 @@
+    int problem_size = 50;
+    args.AddOption(&problem_size, "--problem-size", "--problem-size",
+                   "Scale of problem (number of volumes in x,y directions)");
+
+    // Read upscaling options from command line into upscale_param object
+    upscale_param.RegisterInOptionsParser(args);
+    args.Parse();
+    if (!args.Good())
+    {
+        if (myid == 0)
+        {
+            args.PrintUsage(std::cout);
+        }
+        return 1;
+    }
+    if (myid == 0)
+    {
+        args.PrintOptions(std::cout);
+    }
+    MFEM_ASSERT(fine_samples == 0 ||
+                (coarse_samples == 0 && choose_samples == 0 && shared_samples == 0),
+                "Either do standard (fine) MC, or MLMC, no combination!");
+    serialize["fine-samples"] = picojson::value((double) fine_samples);
+    serialize["coarse-samples"] = picojson::value((double) coarse_samples);
+    serialize["choose-samples"] = picojson::value((double) choose_samples);
+    serialize["shared-samples"] = picojson::value((double) shared_samples);
+    serialize["kappa"] = picojson::value(kappa);
+    // serialize["coarsening-factor"] = picojson::value((double) coarsening_factor);
+
+    // Setting up a mesh
+    std::unique_ptr<mfem::ParMesh> pmesh;
+    if (dimension == 3)
+    {
+        mfem::Mesh mesh(problem_size, problem_size, 10, mfem::Element::HEXAHEDRON, 1,
+                        problem_size * 100.0, problem_size * 100.0, 100.0);
+        pmesh = make_unique<mfem::ParMesh>(comm, mesh);
+    }
+    else
+    {
+        mfem::Mesh mesh(problem_size, problem_size, mfem::Element::QUADRILATERAL, 1,
+                        problem_size * 100.0, problem_size * 100.0);
+        pmesh = make_unique<mfem::ParMesh>(comm, mesh);
+    }
+
+    // Construct a graph from a finite volume problem defined on the mesh
+    mfem::Array<int> ess_attr(dimension == 3 ? 6 : 4);
+    ess_attr = 0;
+    ess_attr[0] = 1;
+    DarcyProblem fvproblem(*pmesh, ess_attr);
+    double cell_volume = fvproblem.CellVolume();
+    Graph graph = fvproblem.GetFVGraph();
+    Upscale upscale(graph, upscale_param, nullptr, &ess_attr);
+
+    upscale.PrintInfo();
+    upscale.ShowSetupTime();
+    upscale.MakeSolver(0);
+
+    mfem::BlockVector rhs_fine(upscale.GetBlockVector(0));
+    rhs_fine.GetBlock(0) = 0.0;
+    rhs_fine.GetBlock(1) = 0.0;
+    // it may make sense to have a more general, not hard-coded, right-hand side
+    int index = 0.5 * problem_size * problem_size + 0.1 * problem_size;
+    rhs_fine.GetBlock(1)(index) = 1.0;
+    index = 0.6 * problem_size * problem_size + 0.75 * problem_size;
+    rhs_fine.GetBlock(1)(index) = -1.0;
+
+    const int seed = argseed + myid;
+    PDESampler sampler(dimension, cell_volume, kappa, seed,
+                       graph, upscale_param, nullptr, &ess_attr);
+
+    mfem::Vector functional(rhs_fine.GetBlock(1));
+    functional = 0.0;
+    double dscale = 1.0 / problem_size;
+    for (int i = 0; i < problem_size; ++i)
```

Comment:
My understanding is that your are defining a functional to compute some sort of average pressure on the bottom boundary (as stated in the description). This definition is not computing that in 3D? But we can measure anything, so it doesn't really matter, it's up to you whether to change this or not.